src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,128 @@
-
+        checkIndex(index);
+        if (size == currentCapacity) {
+            resize();
+        }
+        copy(index, ADD_ID);
+        storage[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        if (index == size) {
+            throw new ArrayListIndexOutOfBoundsException("You have passed incorrect index,"
+                    + "index must be smaller then size, actual index == size");
+        }
+        return storage[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        if (index < size) {
+            storage[index] = value;
+            return;
+        }
+        throw new ArrayListIndexOutOfBoundsException("You have passed incorrect index,"
+                + "index must be smaller then size, actual index == size");
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        if (index == size) {
+            throw new ArrayListIndexOutOfBoundsException("You have passed incorrect index,"
+                    + "index must be smaller then size, actual index == size");
+        }
+        T removed = storage[index];
+        copy(index, REMOVE_ID);
+        size--;
+        return removed;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (storage[i] == element || (storage[i] != null && storage[i].equals(element))) {
+                index = i;
+                break;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException("We cant remove element which doesn't "
+                    + "exist in the list");
+        }
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void resize() {
+        T[] temporary = (T[]) new Object[currentCapacity];
+        System.arraycopy(storage, 0, temporary, 0, size);
+        storage = (T[]) new Object[(int) (currentCapacity * 1.5)];
+        System.arraycopy(temporary, 0, storage, 0, currentCapacity);
+        currentCapacity = (int) (currentCapacity * 1.5);
+    }
+
+    private void copy(int index, int operation) {
+        if (operation == ADD_ID) {
+            System.arraycopy(storage, index, storage, index + 1, size - index);
+        }
+        if (operation == REMOVE_ID) {
+            System.arraycopy(storage, index + 1, storage, index, size - index - 1);
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index > size || index < 0) {
```

Comment:
```suggestion
        if (index >= size || index < 0) {
```
if you write this conditional and you can  use this method in 3 cases (without duplicate cal error when index == size)